core/src/main/java/com/yahoo/ycsb/TimeseriesDB.java

```diff
@@ -0,0 +1,282 @@
+          // FIXME instead return BAD_REQUEST?
+          System.err.print("Downsampling specification for Scan did not match configured downsampling");
+        }
+      } else {
+        String[] queryParts = field.split(tagPairDelimiter);
+        if (queryParts.length == 1) {
+          // we should probably warn about this being ignored...
+          System.err.println("Grouping by arbitrary series is currently not supported");
+          groupByFields.add(field);
+        } else {
+          tagQueries.computeIfAbsent(queryParts[0], k -> new ArrayList<>()).add(queryParts[1]);
+        }
+      }
+    }
+    return scan(table, start, end, tagQueries, downsamplingFunction, downsamplingInterval, timestampUnit);
+  }
+
+  /**
+   * Perform a range scan for a set of records in the database. Each value from the result will be stored in a
+   * HashMap.
+   *
+   * @param metric    The name of the metric
+   * @param startTs   The timestamp of the first record to read.
+   * @param endTs     The timestamp of the last record to read.
+   * @param tags      actual tags that were want to receive (can be empty).
+   * @param aggreg    The aggregation operation to perform.
+   * @param timeValue value for timeUnit for aggregation
+   * @param timeUnit  timeUnit for aggregation
+   * @return A {@link Status} detailing the outcome of the scan operation.
+   */
+  protected abstract Status scan(String metric, Long startTs, Long endTs, Map<String, List<String>> tags,
+                                 AggregationOperation aggreg, int timeValue, TimeUnit timeUnit);
+
+  @Override
+  public Status update(String table, String key, Map<String, ByteIterator> values) {
+    return Status.NOT_IMPLEMENTED;
+    // not supportable for general TSDBs
+    // can be explicitly overwritten in inheriting classes
+  }
+
+  @Override
+  public final Status insert(String table, String key, Map<String, ByteIterator> values) {
+    NumericByteIterator tsContainer = (NumericByteIterator) values.remove(timestampKey);
+    NumericByteIterator valueContainer = (NumericByteIterator) values.remove(valueKey);
+    if (!valueContainer.isFloatingPoint()) {
+      // non-double values are not supported by the adapter
+      return Status.BAD_REQUEST;
+    }
+    return insert(table, tsContainer.getLong(), valueContainer.getDouble(), values);
+  }
+
+  /**
+   * Insert a record in the database. Any tags/tagvalue pairs in the specified tags HashMap and the given value
+   * will be written into the record with the specified timestamp
+   *
+   * @param metric    The name of the metric
+   * @param timestamp The timestamp of the record to insert.
+   * @param value     actual value to insert
+   * @param tags      A HashMap of tag/tagvalue pairs to insert as tagsmv c
+   * @return A {@link Status} detailing the outcome of the insert
+   */
+  protected abstract Status insert(String metric, Long timestamp, double value, Map<String, ByteIterator> tags);
```

Comment:
`Long` -> `long`. And I know a lot of TSDs only support doubles (grrr) but we should have an override for signed ints too. Implementations can return a `Status.NOT_IMPLEMENTED` if they don't support storing ints.